src/domain/navigation/Navigation.ts

```diff
@@ -16,27 +16,39 @@ limitations under the License.
 
 import {BaseObservableValue, ObservableValue} from "../../observable/ObservableValue";
 
-export class Navigation {
-    constructor(allowsChild) {
+
+type AllowsChild<T> = (parent: Segment<T> | undefined, child: Segment<T>) => boolean;
+
+export type OptionalValue<T> = T extends true? [(undefined | true)?]: [T];
```

Comment:
Yeah this is a bit unsightly.
It is used like this (code from Segment ctor):
```js
    constructor(public type: K, ...value: OptionalValue<T[K]>) {
        this.value = (value[0] === undefined ? true : value[0]) as unknown as T[K];
    }
```
The `value` parameter is made a [rest parameter](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-3.html#generic-parameter-defaults) so that we can make it both optional and give it a default value of undefined depending on the segment type.
So for example, these should be accepted:
```js
new Segment("create-room"); // value is optional, internally given the value true
new Segment("create-room", true);
```
where as these should not be accepted:
```js
new Segment("room") // error: second argument is needed and it should be string (room-id)
```
`OptionalValue` is basically stating that if SegmentType[type] = true:
1. Allow this argument to be optional
2. Give it a default value of undefined
3. Also allow it to be true

I tried a lot of things but they would all require that we explicitly pass undefined as second argument, eg:
```js
constructor(public type: K, value: T[K] extends true? undefined: T[K]) {
    this.value = (value[0] === undefined ? true : value[0]) as unknown as T[K];
}

const s: Segment<SegmentType> = new Segment("create-room"); // error: An argument for 'value' was not provided
const s: Segment<SegmentType> = new Segment("create-room", undefined); //works, but not what we want
```

Reply:
right, makes sense! Can you perhaps put all of that in a comment?